app/src/main/java/com/ncnf/database/DatabaseService.java

```diff
@@ -1,125 +1,245 @@
+    @Override
+    public <T> CompletableFuture<T> getField(String documentPath, String field){
+        CompletableFuture<T> futureResponse = new CompletableFuture<>();
 
+        this.db.document(documentPath).get().addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                T result = (T) task.getResult().get(field);
+                futureResponse.complete(result);
+            } else {
+                futureResponse.completeExceptionally(task.getException());
+            }
+        });
         return futureResponse;
     }
 
     @Override
-    public CompletableFuture<DatabaseResponse> updateField(String path, String field, Object value) {
-        CompletableFuture<DatabaseResponse> futureResponse = new CompletableFuture<>();
+    public CompletableFuture<Boolean> updateField(String documentPath, String field, Object value) {
+        CompletableFuture<Boolean> futureResponse = new CompletableFuture<>();
+
+        this.db.document(documentPath).update(FieldPath.of(field), value).addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                futureResponse.complete(true);
+            } else {
+                futureResponse.completeExceptionally(task.getException());
+            }
+        });
 
-        this.db.document(path).update(FieldPath.of(field), value).addOnCompleteListener(task -> onTaskComplete(task, futureResponse));
         return futureResponse;
     }
 
     @Override
-    public CompletableFuture<DatabaseResponse> getField(String path, String field) {
-        return getData(path).thenApply(task -> {
+    public <T> CompletableFuture<T> getDocument(String documentPath, Class<T> documentType){
+        CompletableFuture<T> futureResponse = new CompletableFuture<>();
+
+        this.db.document(documentPath).get().addOnCompleteListener(task -> {
             if(task.isSuccessful()){
-                Object obj = ((Map<String, Object>) task.getResult()).get(field);
-                return new DatabaseResponse(true, obj, task.getException());
+                DocumentSnapshot document = (DocumentSnapshot) task.getResult();
+
+                T result = (T) registry.get(documentType).toObject(document.getId(), document.getData());
+                futureResponse.complete(result);
             } else {
-                return task;
+                futureResponse.completeExceptionally(task.getException());
             }
         });
+
+        return futureResponse;
     }
 
     @Override
-    public CompletableFuture<DatabaseResponse> getData(String path) {
-        CompletableFuture<DatabaseResponse> futureResponse = new CompletableFuture<>();
+    public <T> CompletableFuture<Boolean> setDocument(String documentPath, T object) {
+        CompletableFuture<Boolean> futureResponse = new CompletableFuture<>();
+
+        Map<String, Object> data = registry.get(object.getClass()).toMap(object);
```

Comment:
You're implementation is good and I guess that fact that you can't access a generic's object type at runtime in java forces you to do something like this. I'm no one to judge but great workaround!

Reply:
Yes exactly, I had help from Eloi and Dorian !